File: SortingAlgorithms/MergeSort.java
public class MergeSort{

	MergeSort() {}
	
	public void mSort(int[] arr, int n) {

		if (n < 2) {return;}

		int mid = n/2;
		int left[] = new int[mid];
		int right[] = new int[n - mid];

		for (int i =0; i < mid; i++) {
			left[i] = arr[i];
		}

		for (int i = mid; i < n; i++) {
			right[i - mid] = arr[i];
		}

		mSort(left, mid);
		mSort(right, n - mid);

		merge(arr, left, right, mid, n - mid);
		
	}

	public void merge(int[] arr, int[] left, int[] right, int l, int r) {
		int i = 0;
		int j = 0;
		int k = 0;

		while (i < l && j < r) {
			if (left[i] <= right[j]) {
				arr[k] = left[i];
				k++;
				i++;
			} else {
				arr[k] = right[j];
				k++;
				j++;
			}
		}

		while (i < l) {
			arr[k] = left[i];
			k++;
			i++;
		}
		while (j < r) {
			arr[k] = right[j];
			k++;
			j++;
		}
	}

	
	public static void main(String[] args){
	    int[] arr = { 5, 1, 6, 2, 3, 4 };

	    MergeSort mergeSort = new MergeSort();

		for (int ele: arr) {
			System.out.print(ele + " ");
		}

		System.out.println();
	    mergeSort.mSort(arr, arr.length);
		
		for (int ele: arr) {
			System.out.print(ele + " ");
		}	    
	}
}

File: SortingAlgorithmsImpl/MergeSort.cpp
#include <iostream>

using namespace std;

class MergeSort {

	public:
		void sort(int arr[], int n) {

			int mid = n/2;
			int leftArr[mid];
			int rightArr[n - mid];
			
			if (n < 2) {
				return;
			}

			for (int i = 0; i < mid; i++) {
				leftArr[i] = arr[i];
			}

			for (int j = 0; j < n - mid ; j++) {
				rightArr[j] = arr[mid + j];
			}
			
			sort(leftArr, mid);
			sort(rightArr, n - mid);

			merge(arr, leftArr, rightArr, mid, n -mid);
		}


		void merge(int arr[], int leftArr[], int rightArr[], int leftSize, int rightSize) {
			// i iterator for left arr, j iterator for right arr, k for final arr
			int i = 0, j = 0, k = 0;

			while (i < leftSize && j < rightSize) {

				if (leftArr[i] < rightArr[j]) {
					arr[k++] = leftArr[i++];
				} else {
					arr[k++] = rightArr[j++];
				}
			}

			//In case the left array ran out of room to compare you fill the rest with right arr
			//vice versa for the right array.  
			while (i < leftSize) {
				arr[k++] = leftArr[i++];
			}

			while (j < rightSize) {
				arr[k++] = rightArr[j++];
			}
			
		}	
};

int main() {

	int arr[] = {3, 10, 1, 32, 11, 5};
	int n = sizeof(arr) / sizeof(int); 

	for (int i = 0; i < n; i++) {
		cout << arr[i] << " ";
	}

	cout << "\nAfter merge sort... "<< endl;	
	MergeSort mergeSort;
	mergeSort.sort(arr, n);

	for (int i = 0; i < n; i++) {
		cout << arr[i] << " ";
	}

}
